spec.bs

```diff
@@ -1989,6 +1992,71 @@ finalized.
+navigable container/fenced navigable=].
+
+For any successful navigation inside of a <{fencedframe}>, regardless of whether it was initiated by
+content in the <{fencedframe}> or its embedder, exactly one of three things will happen:
+
+ * <i>Embedder-initiated navigations</i>: The [=navigation params=]'s [=navigation params/fenced
+   frame config instance=] will be unconditionally assigned to the <{fencedframe}>'s [=fenced
+   navigable container=]'s [=fenced navigable container/fenced navigable=]'s [=navigable/active
+   browsing context=]'s [=browsing context/fenced frame config instance=].
+
+ * <i>Cross-origin inner-content-initiated navigations</i>: We do not want the <{fencedframe}>'s
+   [=fenced navigable container=]'s [=fenced navigable container/fenced navigable=]'s
+   [=navigable/active browsing context=]'s [=browsing context/fenced frame config instance=] to
+   persist through this kind of cross-{{Document}} same-[=browsing context group=] navigation, so
+   we manually clear it.
+
+ * <i>Same-origin inner-content-initiated navigations</i>: The [=fenced frame config instance=]
+   referenced in the immediately-preceding point persists through this kind of cross-{{Document}}
+   navigation. The patches below take no action to clear it, so it continues to be visible through
+   the {{Window/fence}} API.
+
+<div algorithm=create-and-initialize-document-patch>
+  Modify [[!HTML]]'s [=create and initialize a Document object=] algorithm to insert a step after
+  step 2 that reads:
+
+  3. If |navigationParams|'s [=navigation params/fenced frame config instance=] is not null:
+
+    1. [=Assert=]: |browsingContext| does not equal |navigationParams|'s [=navigation
+       params/navigable=]'s [=navigable/active browsing context=].
+
+       Note: This is because embedder-initiated navigations (indicated by |navigationParams|'s
+       [=navigation params/fenced frame config instance=] being non-null) always cause a
+       [[#bcg-swap]].
+
+    1. Set |browsingContext|'s [=browsing context/fenced frame config instance=] to
+       |navigationParams|'s [=navigation params/fenced frame config instance=].
+
+  4. Otherwise, if |navigationParams|'s [=navigable/active browsing context=]'s [=browsing
+     context/fenced frame config instance=] is non-null, and either of the following conditions are
+     true:
+
+       * |navigationParams|'s [=navigation params/origin=] is not [=same origin=] with
+         |navigationParams|'s [=navigation params/navigable=]'s [=navigable/active document=]'s
+         [=Document/origin=];
+
+       * |navigationParams|'s [=navigation params/response=]'s
+         [=response/has-cross-origin-redirects=] boolean is true
+
+     Then set |navigationParams|'s [=navigable/active browsing context=]'s [=browsing context/fenced
+     frame config instance=] to null.
+
+     Note: This ensures that the {{Window/fence}} API return null for script running in the
+     {{Document}} created by this algorithm.
```

Comment:
@gtanzer I can't remember if this is the goal (to completely clear `window.fence`) or if we just want to clear *parts* of the config, so please let me know if this does not match the implementation.

Also if you could point me towards any WPTs demonstrating this behavior it would be great to link to them from here.

Reply:
Liam recently changed it to clear parts of the config instance, but that was just motivated by the way permissions work in impl across processes. I don't think the spec needs to work like that.

The config instance semantics are:
* Embedder-initiated navigations unconditionally set the config instance for the fenced frame tree
* Future accesses to the config instance through window.fence APIs, etc. are gated by a check on whether the accessing origin is same-origin with the config instance's mapped url.

I'm not sure whether the latter bullet really should be spec'd as nulling the config instance. We still want the information associated with it.

The spec for reportEvent (https://wicg.github.io/fenced-frame/#dom-fence-reportevent) and other APIs already include this origin check.

Reply:
> We still want the information associated with it.

Can you elaborate? I'm not sure what "information" is here, but it sounds like for cross-origin-to-config-instance iframes within fenced frames (or even fenced frames that have navigated cross-origin to their mapped URL at some point), the `window.fence` API should still be non-null and the APIs on it should still be exposed, its just that they should each decide whether to work or not based on the current origin. Is that right?

Reply:
Right, each API should decide whether or not to work based on the current origin. The fact that cross-origin frames "don't have" a config instance in the renderer implementation is just so compromised renderers can't see that information.

At a spec level, you just have a config instance that governs the entire fenced frame tree and ad hoc checks for each API that uses the data in it.